presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1104,34 +1107,32 @@ else if (criteria instanceof JoinOn) {
 
             for (Expression conjunct : ExpressionUtils.extractConjuncts((Expression) optimizedExpression)) {
                 conjunct = ExpressionUtils.normalize(conjunct);
```

Comment:
Ya, lets not make major changes to the planning of JOIN in this PR. It
shouldn't require adding the original predicate to the JoinNode when we do
though. It can be planned as a join followed by a filter, and rely on
predicate push down.

I think the normalization of not not a.x = b.x can be done pretty easily in
the optimizers, in the way I mentioned earlier

On Thu, Dec 31, 2015, 9:02 AM Łukasz Osipiuk <notifications@github.com>
wrote:

> In
> presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java
> <https://github.com/facebook/presto/pull/3954#discussion_r48661167>:
>
> > @@ -1104,34 +1107,32 @@ else if (criteria instanceof JoinOn) {
> >
> >              for (Expression conjunct : ExpressionUtils.extractConjuncts((Expression) optimizedExpression)) {
> >                  conjunct = ExpressionUtils.normalize(conjunct);
>
> What you write makes sense. Though if, I understand correctly changing the
> code the way you suggest, is total rebuild of how join is planned today. We
> would need to add original join condition into JoinNode (it is not there
> right now - only list of EquiJoinClauses). And then move most of the
> logic from RelationPlanner.visitJoin to one or more optimizers.
>
> As currently all the detailed planning happens in
> RelationPlanner.visitJoin I cannot move the normalization into
> SimplifyExpressions as this one is executing *after*
> RelationPlanner.visitJoin.
>
> I would rather not put this Join planning refactor as part of this PR.
> Honestly I am not sure if with my current knowledge I would be able to
> efficiently implement this at all :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/presto/pull/3954/files#r48661167>.
>